Mathlib/Tactic/Linter/UnusedSetOptionIn.lean

```diff
@@ -0,0 +1,65 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.Tactic.Lemma
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+The `unusedSetOptionIn` linter warns against "most external" unused `set_option ... in`s.
+-/
+
+open Lean Elab Command
+
+/-- converts
+* `theorem x ...` to  `some (example ... , x)`,
+* `lemma x ...`   to  `some (example ... , x)`,
+* `example ...`   to ``some (example ... , `example)``,
+*  everything else goes to `none`.
+-/
+def toExample {m : Type → Type} [Monad m] [MonadRef m] [MonadQuotation m] :
+    Syntax → m (Option (Syntax × Syntax))
+  | `($dm:declModifiers theorem $did:declId $ds* : $t $dv:declVal) => do
+    return some (← `($dm:declModifiers example $ds* : $t $dv:declVal), did.raw[0])
+  | `($dm:declModifiers lemma $did:declId $ds* : $t $dv:declVal) => do
+    return some (← `($dm:declModifiers example $ds* : $t $dv:declVal), did.raw[0])
+  | `($dm:declModifiers example $ds:optDeclSig $dv:declVal) => do
+    return some (← `($dm:declModifiers example $ds $dv:declVal), mkIdent `example)
+  | _ => return none
+
+/-- The `unusedSetOptionIn` linter warns against "most external" unused `set_option ... in`s. -/
+register_option linter.unusedSetOptionIn : Bool := {
+  defValue := true
+  descr := "enable the unusedSetOptionIn linter"
+}
+
+/-- Gets the value of the `linter.unusedSetOptionIn` option. -/
+def getSetOptionIn (o : Options) : Bool := Linter.getLinterValue linter.unusedSetOptionIn o
+
+/-- reports a warning if the "first layer" `set_option ... in` is unused. -/
+def findSetOptionIn (cmd : CommandElab) : CommandElab := fun stx => do
+  let mut report? := (false, default)
+  let s ← get
+  match stx with
+    | .node _ ``Lean.Parser.Command.in #[
+        .node _ ``Lean.Parser.Command.set_option _,
```

Comment:
Do you want to differentiate which option is set? For instance, if I set `profiler.threshold 50 in`, I want to have this for temporary purposes - and linting on it seems more dubious than useful.
(The `set_option` linter also covers that case - written in Python, but its rewrite in Lean is awaiting review.)

Reply:
I was planning to discriminate on which option is being set by going over what the lniter flags.  For instance, I am not sure whether this linter should handle various thresholds/heartBeats, since maybe a more dedicated one could also do some diagnostics on the side.